initial commit for pkg iceplume

antnguyen13 opened the pull request:
## What changes does this PR introduce?
Introduce a new package "iceplume" to parameterize melt plumes at glacier fronts as well as code in main branch that propagate the properties computed from the pkg to state variables.


## What is the current behaviour? 
There is no current behavior. This commit submit a new pkg.


## What is the new behaviour 
A new package to parameterize melt plumes at glacier fronts


## Does this PR introduce a breaking change? 
It should not, because it is a new package.

## Other information:
Reference: Cowton, T., D. Slater, A. Sole, D. Goldberg, and P. Nienow (2015), Modeling the impact of glacial runoff on fjord circulation and submarine melt rate using a new subgrid-scale parameterization for glacial plumes, J. Geophys. Res. Oceans, 120, 796–812, doi:10.1002/2014JC010324.

## Suggested addition to `tag-index`

antnguyen13 commented:
@mjlosch @jm-c 
not sure if i did it right... but the test case is here in my repo, i put it inside verification (!)
https://github.com/antnguyen13/MITgcm/tree/iceplume_testcase/verification/iceplume_testcase

ifenty commented:
@antnguyen13 So glad that you've done this initial PR

mjlosch commented:
@antnguyen13 I am not sure what to do here. The test contains 8 code directories, the "results" directory correspondingly 8 STDOUT files (but no output.txt or similar), and 8 Makefiles (?), the provided script is specific to a specific machine that I don't have access to. So I cannot use testreport.

Instead, I added `iceplume` in `isomip/code/packages.conf` just to see if it compiles and I immediately get compilation errors starting in iceplume_calc/iceplume_plume_model, etc., (many undefined variables) and many "nonconforming tab character" warnings,  "Fortran 2018 deleted feature: ..." warnings from the pkg (and compilation errors, no implicit type etc.). Does this package only work with very specific setups (to be specified in the code directory?)

I noticed that every single line of PARAMS.h changes in this PR, why?

I'll wait until this is a little easier to test.

antnguyen13 commented:
@mjlosch I think because I checked in ALL my mods already in the PR, likely the only one that will work is code_mod7 if you use the PR /pkg/iceplume.  The example I provided is for a specific configuration Tom provided, a box of certain size, with masks that he provided in input_binary.  The .bash script is self-contained i believe: in that if you enter mod=7, the execute the mod, AFTER YOU MODIFY the Makefile to point to your own optfile for your machine (replace -of=/home/atnguyen/nansen/computing/optfiles/optfile_mvapich2_sverdrup with yours, or leave as default?) .

It is true that I think due to the solvers in his provided code has some specific nonconforming syntax that you pointed out, I had to fiddle around with the opt file on our side to get 1 version to compile .  Let me see if I can update it to the branch.

antnguyen13 commented:
@mjlosch  "noticed that every single line of PARAMS.h changes in this PR, why?"

I'm not sure what this means.  Could it be some returning/end-of-line character have been inserted?  I'm on a linux machine so that shouldn't have been the problem.. but i'd like to understand more what this is.

mjlosch commented:
Have a look at "files changed" or do a "git diff upstream/master" in your repository. All lines in PARAMS.h have been removed and then replaced by apparently the same lines. I have no idea, how that can happen.

antnguyen13 commented:
"Have a look at "files changed" or do a "git diff upstream/master" in your repository. All lines in PARAMS.h have been removed and then replaced by apparently the same lines. I have no idea, how that can happen."
@mjlosch  oh, indeed i see that now in the diff, should I make an attempt to point it to track upstream, and then just checkout that upstream PARAMS.h to get rid of these diff? (not sure how to do yet.. will ask some help locally here..)

antnguyen13 commented:
@mjlosch : thanks to @jahn 's help, i fixed some outstanding issues related to exceeding 72char in array definitions in ICEPLUME.h , and also now added implicit none and defined variables in the solvers, such that my simple adding of iceplume into isomip/code compiles on my side with default optfile linux_amd64_gfortran

antnguyen13 commented:
@jm-c & @mjlosch : i've created a new PR in verification_other, https://github.com/MITgcm/verification_other/pull/58 .  Once this set up is copied in verification, or put into MITgcm/verification_other/ level , it should compile and run successfully with this current PR 652
